### DISCUSSION TITLE
Github Pull Request Description Enforcement

### DISCUSSION THREAD

**Discussion by karkir0003** [community/community#84771] [Pull Requests]
### Select Topic Area

General

### Body

Hey team 👋 . Hope everyone had a Happy New Year! I'm leading a team of 10+ developers on an independent full stack project. I'm able to configure a pull request template using these [docs](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository). However, I noticed that everytime a pull request is created, the markdown file is populated with the template. While this is great, I've noticed that there is a lack of enforcement on my team following the PR structure. 

I'm wondering how I can configure the pull request tab to resemble more of a form based structure with display of the github action build checks and stuff running. With a form based structure (similar to issues/bug templates), I can better enforce good PR description structure, but I'd still like developers to be able to see the GitHub actions running to ensure the build checks passed. 

Can this be configured in the current state of GitHub? If so, could you point me to sample repo/docs? If not, could the GitHub team please consider implementing this feature to boost developer productivity and improve on PR standardization (this has opportunity on business/enterprise + hobby use cases)

TLDR of the question: Similar to how we can configure issue/bug report form templates in the Issues tab, can I configure a form-based template for Pull Requests to enforce good PR description structure

**Comment by bryan-shea** [comment]
### Hello, @karkir0003 

Happy holidays! 

As of now, GitHub does not support form-based templates for pull requests. PR templates in GitHub are markdown files that serve as a guide but do not enforce a structured format like issue templates.

### Solutions / Workarounds:

1. **Use Markdown Templates Effectively:**
   - Clearly outline the expected structure in your markdown PR template.
   - Include a checklist within the template to guide developers.

2. **Enforce Standards with GitHub Actions:**
   - Implement a GitHub Action to check the PR description for compliance with the template.
   - The action can fail the PR if it doesn't meet the specified criteria, necessitating edits for compliance.

3. **Team Practices:**
   - Educate and regularly remind your team about the importance of following the PR template.
   - Conduct PR reviews focusing on adherence to the template.

### Outline for GitHub Action Implementation:

1. **Trigger:** Set the action to run on pull request events (`opened`, `edited`, `reopened`).
2. **Validation Job:** Create a job that validates the PR description against the template.
3. **Failure Condition:** If the PR description does not match the template, the job fails, blocking the PR merge.

### Sample GitHub Action YAML (Outline):

```yaml
name: PR Description Validation

on:
  pull_request:
    types: [opened, edited, reopened]

jobs:
  validate-pr-description:
    runs-on: ubuntu-latest
    steps:
      - name: Check PR Description Against Template
        run: |
          # Script to validate PR description
          # Fail the job if it doesn't adhere to the template
```

While GitHub currently does not offer form-based PR templates, a combination of markdown PR templates, GitHub Actions for validation, and team training can effectively enforce PR description standards. For future updates, consider submitting a feature request to GitHub.

**Comment by karkir0003** [comment]
> ### Hello, @karkir0003 
> 
> Happy holidays! 
> 
> As of now, GitHub does not support form-based templates for pull requests. PR templates in GitHub are markdown files that serve as a guide but do not enforce a structured format like issue templates.
> 
> ### Solutions / Workarounds:
> 
> 1. **Use Markdown Templates Effectively:**
>    - Clearly outline the expected structure in your markdown PR template.
>    - Include a checklist within the template to guide developers.
> 
> 2. **Enforce Standards with GitHub Actions:**
>    - Implement a GitHub Action to check the PR description for compliance with the template.
>    - The action can fail the PR if it doesn't meet the specified criteria, necessitating edits for compliance.
> 
> 3. **Team Practices:**
>    - Educate and regularly remind your team about the importance of following the PR template.
>    - Conduct PR reviews focusing on adherence to the template.
> 
> ### Outline for GitHub Action Implementation:
> 
> 1. **Trigger:** Set the action to run on pull request events (`opened`, `edited`, `reopened`).
> 2. **Validation Job:** Create a job that validates the PR description against the template.
> 3. **Failure Condition:** If the PR description does not match the template, the job fails, blocking the PR merge.
> 
> ### Sample GitHub Action YAML (Outline):
> 
> ```yaml
> name: PR Description Validation
> 
> on:
>   pull_request:
>     types: [opened, edited, reopened]
> 
> jobs:
>   validate-pr-description:
>     runs-on: ubuntu-latest
>     steps:
>       - name: Check PR Description Against Template
>         run: |
>           # Script to validate PR description
>           # Fail the job if it doesn't adhere to the template
> ```
> 
> While GitHub currently does not offer form-based PR templates, a combination of markdown PR templates, GitHub Actions for validation, and team training can effectively enforce PR description standards. For future updates, consider submitting a feature request to GitHub.

Thank you for the detailed response. Can you link a sample working/functional github action script that follows the structure you outlined for validating pr description format/convention

**Reply by bryan-shea** [reply]
Of course, I'll include both a video playlist that goes a bit more into GitHub actions along with a working example script for you down below! Please feel free to modify it for your specific use case/needs or message me if you need any further help! 

#### Video Link & Useful Repos:
[GitHub Actions: Full Course](https://youtube.com/playlist?list=PLArH6NjfKsUhvGHrpag7SuPumMzQRhUKY&si=eueXHGjzSz_BzO9d)
[GitHub Script](https://github.com/actions/github-script)
[Script Example for above repo](https://github.com/actions/github-script/blob/main/action.yml)

---

#### Working Script Example: 

```yaml

name: PR Description Validator

on:
  pull_request:
    types: [opened, edited, reopened]

jobs:
  validate-pr-description:
    name: Validate PR Description
    runs-on: ubuntu-latest
    steps:
    - name: Set up workspace
      uses: actions/checkout@v2

    - name: Validate PR Description
      run: |
        # Ensure jq is installed
        sudo apt-get install jq

        # Fetch PR description using jq
        PR_DESCRIPTION=$(jq -r ".pull_request.body" "$GITHUB_EVENT_PATH")
        KEYWORD="REQUIRED_KEYWORD"

        # Check for the required keyword in the PR description
        if [[ $PR_DESCRIPTION != *"$KEYWORD"* ]]; then
          echo "PR description does not contain the required keyword."
          exit 1
        fi
```

"In the revised script:

- The action is triggered by `pull_request` events, specifically when a PR is opened, edited, or reopened. This ensures that the validation check is performed at every critical stage of the PR's lifecycle.
- The `validate-pr-description` job begins by setting up the workspace using `actions/checkout@v2`. This step is essential for scripts that might need access to the code in the repository.
- The validation script, included in the job, is a Bash script that first ensures the `jq` tool is installed. `jq` is used for parsing JSON data, which is necessary to extract the PR description from the GitHub event payload.
- The script then checks the PR description for a specific keyword, denoted as `REQUIRED_KEYWORD` in the script. This is a basic form of validation; you can replace or expand this logic to suit your specific validation criteria.
- If the PR description does not meet the set criteria (such as containing the required keyword), the script exits with a status of 1. This exit status signals a failure in the validation step, causing the GitHub Action to fail. This failure acts as an indicator that the PR description does not adhere to the project's guidelines.

While this example is still basic, it includes an installation step for a necessary tool (`jq`) and serves as a more practical starting point. For more intricate validation requirements, you may consider enhancing the script with additional logic.
